rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,479 @@
+/// 4. Another thought is to leave as is, and write a crate on top of this one, called something like symcrypt_rsa_parse, which will be a friendlier interface on top of the RSA code we have right now
+/// It would have convenience functions like rsa_from_pkcs8, rsa_from_der, rsa_from_asn1 etc. ( this could be part of the intern project as well.)
+/// 5. with how many steps / options there is for generating a key, we can maybe do a builder pattern, this would be similar to how it is on NCrypt / BCrypt where they have 
+/// RsaKey functions and the RsaKey.finalize() at the end.
+/// ex:
+///         builder = RsaKeyBuilder::new()
+///             .key_mode(RsaKeyMode::KeyPair)
+///             .key_usage(RsaKeyUsage::Sign)
+///             .generate_key()
+
+
+/// Impl for the RsaKeyPair
+impl RsaKeyPair {
+    /// `new()` allocates a the space needed for an `RsaKeyPair`.
+    /// To fill data, you must call one of [`RsaKeyPair::generate_key()`], [`RsaKeyPair::set_public_only()`] or [`RsaKeyPair::set_key_pair()`].
+    /// 
+    /// `n_bits_mod` represents the desired bit length of the Rsa Key.
+    /// 
+    /// `key_mode` will determine if this key will be allocated for a [`RsaKeyMode::PublicOnly`] or [`RsaKeyMode::KeyPair`] 
+    /// 
+    /// `key_usage` will indicate if this key will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`]
+    pub fn new(n_bits_mod: u32, key_mode: RsaKeyMode, key_usage: RsaKeyUsage) -> Self {
+        let rsa_params = symcrypt_sys::SYMCRYPT_RSA_PARAMS {
+            version: 1 as symcrypt_sys::UINT32, // No other version aside from version 1 is specified
+            nBitsOfModulus: n_bits_mod as symcrypt_sys::UINT32,
+            nPrimes: key_mode.to_u32(), //  Value must be 0 or 2, if only a public key is being generated, you can pass 0 for n_primes
+            nPubExp: MAX_NUMBER_OF_PUBLIC_EXPONENTS, // Per SymCrypt documentation, only 1 is allowed
+        };
+        unsafe {
+            // SAFETY: FFI calls
+            // No flags specified for SymCryptRsakeyAllocate
+            let rsa_key = Box::pin(symcrypt_sys::SymCryptRsakeyAllocate(&rsa_params, 0));
+            RsaKeyPair {
+                inner: rsa_key,
+                key_mode: key_mode,
+                key_usage: key_usage,
+                num_p_exp: MAX_NUMBER_OF_PUBLIC_EXPONENTS,
+            }
+        }
+    }
+
+    /// `generate_key()` generates a random Rsa key based on the parameters passed into [`RsaKeyPair::new()`].
+    /// 
+    /// `generate_key()` can only be called after space for the key has been allocated through [`RsaKeyPair::new()`].
+    /// 
+    /// `pub_exp` takes in an `Option<u64>` that is the public exponent. If `None` is provided, the default `2^16 +1` will be used.
+    pub fn generate_key(&mut self, pub_exp: Option<u64>) -> Result<(), SymCryptError> {
```

Comment:
>generate_key [](http://example.com/codeflow?start=11&length=12)

Can this be called more than once?

Reply:
Yes it can, to my understanding this is not a common pattern, see my !Review comment where I bring up the idea of combining the `new()` with the `generate()` 

> 2. The idea of new() then generate_key() / generate_set_public_only() / set_key_pair() is a bit convoluted IMO, we can squash these together to something like: 
 new_random(), and  new_set_public() and new_set_key_pair().